common/transport/amqp/src/amqp_cbs.ts

```diff
@@ -0,0 +1,323 @@
+            let tokenOperation = this._putTokenQueue.shift();
+            while (tokenOperation) {
+              tokenOperation.callback(err);
+              tokenOperation = this._putTokenQueue.shift();
+            }
+
+            /*Codes_SRS_NODE_AMQP_CBS_16_006: [If given as an argument, the `attach` method shall call `callback` with a standard `Error` object if any link fails to attach.]*/
+            if (callback) {
+              callback(err);
+            }
+          },
+          attach: (callback) => {
+            this._fsm.transition('attaching', callback);
+          },
+          detach: (callback) => { if (callback) callback(); },
+          forceDetach: () => { return; },
+          putToken: (audience, token, callback) => {
+            this._putTokenQueue.push({
+              audience: audience,
+              token: token,
+              callback: callback
+            });
+            this._fsm.transition('attaching');
+          }
+         },
+        attaching: {
+          _onEnter: (callback) => {
+            this._senderLink.attach((err) => {
+              if (err) {
+                this._fsm.transition('detached', callback, err);
+              } else {
+                this._receiverLink.attach((err) => {
+                  if (err) {
+                    this._fsm.transition('detaching', callback, err);
+                  } else {
+                    this._receiverLink.on('message', (msg) => {
+                      //
+                      // Regardless of whether we found the put token in the list of outstanding
+                      // operations, accept it.  This could be a put token that we previously
+                      // timed out.  Be happy.  It made it home, just too late to be useful.
+                      //
+                      /*Codes_SRS_NODE_AMQP_CBS_16_020: [All responses shall be completed.]*/
+                      this._receiverLink.accept(msg);
+                      for (let i = 0; i < this._putToken.outstandingPutTokens.length; i++) {
+                        if (msg.correlationId === this._putToken.outstandingPutTokens[i].correlationId) {
+                          const completedPutToken = this._putToken.outstandingPutTokens[i];
+                          this._putToken.outstandingPutTokens.splice(i, 1);
+                          if (completedPutToken.putTokenCallback) {
+                            /*Codes_SRS_NODE_AMQP_CBS_16_019: [A put token response of 200 will invoke `putTokenCallback` with null parameters.]*/
+                            let error = null;
+                            if (msg.properties.getValue('status-code') !== 200) {
+                              /*Codes_SRS_NODE_AMQP_CBS_16_018: [A put token response not equal to 200 will invoke `putTokenCallback` with an error object of UnauthorizedError.]*/
+                              error = new errors.UnauthorizedError(msg.properties.getValue('status-description'));
```

Comment:
>error = new errors.UnauthorizedError(msg.properties.getValue('status-description')); [](start = 30, length = 84)

This will still transition into attached.  Is that what you want?

Reply:
technically still attached.

---
In reply to: [139537583](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139537583) [](ancestors = 139537583)